libpromises/evalfunction.c

```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
     return vardata;
 }
 
-static Rlist *GetHostsFromLastseenDB(Item *addresses, time_t horizon, bool return_address, bool return_recent)
+/*********************************************************************/
+
+static FnCallResult FnCallAnd(EvalContext *ctx,
+                              ARG_UNUSED const Policy *policy,
+                              ARG_UNUSED const FnCall *fp,
+                              const Rlist *finalargs)
+{
+
+    for (const Rlist *arg = finalargs; arg; arg = arg->next)
+    {
+        SyntaxTypeMatch err = CheckConstraintTypeMatch(fp->name, arg->val, CF_DATA_TYPE_STRING, "", 1);
+        if (err != SYNTAX_TYPE_MATCH_OK && err != SYNTAX_TYPE_MATCH_ERROR_UNEXPANDED)
+        {
+            FatalError(ctx, "Function '%s', %s", fp->name, SyntaxTypeMatchToString(err));
+        }
+    }
+
+    for (const Rlist *arg = finalargs; arg; arg = arg->next)
+    {
+        if (!IsDefinedClass(ctx, RlistScalarValue(arg)))
+        {
+            return FnReturnContext(false);
+        }
+    }
+
+    return FnReturnContext(true);
+}
+
+/************** FnCallHostsSeen utilities **************/
+
+typedef struct {
+    char *address;
+    char *hostkey;
+    char *lastseen;
```

Comment:
Why not use the proper type?

Reply:
This links to the sprintf/sscanf comments bellow. I am not sure why this conversion has to happen too. But the [if check](https://github.com/cfengine/core/blob/master/libpromises/evalfunction.c#L522-L526) implies that something can go wrong but it looks like these sprintf/sscanf conversions are just asking for trouble.

Woudlnt it be ok to just have a `time_t` field and remove that if entirely?